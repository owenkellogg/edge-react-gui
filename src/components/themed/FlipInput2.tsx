import { eq } from 'biggystring'
import React, { useEffect } from 'react'
import { Platform, TextInput, TouchableWithoutFeedback, View } from 'react-native'
import Animated, { AnimationCallback, Easing, interpolate, runOnJS, useAnimatedStyle, useSharedValue, withTiming } from 'react-native-reanimated'

import { Fontello } from '../../assets/vector'
import { useHandler } from '../../hooks/useHandler'
import { formatNumberInput, isValidInput } from '../../locales/intl'
import s from '../../locales/strings'
import { useState } from '../../types/reactHooks'
import { NumericInput } from '../modals/NumericInput'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { EdgeText } from './EdgeText'
import { ButtonBox } from './ThemedButtons'

export interface FlipInputGetMethodsResponse {
  setAmounts: (value: string[]) => void
}

export type FlipInputFieldInfo = {
  currencyName: string

  // Maximum number of decimals to allow the user to enter. FlipInput will automatically truncate use input to this
  // number of decimals as the user types.
  maxEntryDecimals: number
}

export interface FlipInputProps {
  onNext?: () => void
  convertValue: (sourceFieldNum: number, value: string) => Promise<string | undefined>
  getMethods?: (methods: FlipInputGetMethodsResponse) => void
  startAmounts: string[]
  inputAccessoryViewID?: string
  fieldInfos: FlipInputFieldInfo[]
  topReturnKeyType?: 'done' | 'go' | 'next' | 'search' | 'send'
}

const FLIP_DURATION = 500

export const FlipInput2 = React.memo((props: FlipInputProps) => {
  const theme = useTheme()
  const styles = getStyles(theme)
  const inputRefs = [React.useRef<TextInput>(null), React.useRef<TextInput>(null)]

  const { startAmounts, fieldInfos, topReturnKeyType = 'done', onNext, inputAccessoryViewID, getMethods, convertValue } = props
  const animatedValue = useSharedValue(0)

  // `amounts` is always a 2-tuple
  const [amounts, setAmounts] = useState<string[]>(startAmounts)

  // primaryField is the index into the 2-tuple, 0 or 1
  const [primaryField, setPrimaryField] = useState<number>(0)

  const frontAnimatedStyle = useAnimatedStyle(() => {
    const degrees = interpolate(animatedValue.value, [0, 0.5, 1], [0, 90, 90])
    return {
      transform: [{ rotateX: `${degrees}deg` }]
    }
  })
  const backAnimatedStyle = useAnimatedStyle(() => {
    const degrees = interpolate(animatedValue.value, [0, 0.5, 1], [90, 90, 0])
    return {
      transform: [{ rotateX: `${degrees}deg` }]
    }
  })

  const onToggleFlipInput = useHandler(() => {
    const jsCallback: AnimationCallback = done => {
      'worklet'
      if (done) runOnJS(setPrimaryField)(Number(!primaryField))
    }
    inputRefs[primaryField].current?.blur()
    inputRefs[Number(!primaryField)].current?.focus()

    if (primaryField) {
      console.log('animating to 0')
      animatedValue.value = withTiming(
        0,
        {
          duration: FLIP_DURATION,
          easing: Easing.inOut(Easing.ease)
        },
        jsCallback
      )
    }
    if (!primaryField) {
      console.log('animating to 1')
      animatedValue.value = withTiming(
        1,
        {
          duration: FLIP_DURATION,
          easing: Easing.inOut(Easing.ease)
        },
        jsCallback
      )
    }
  })

  const onNumericInputChange = useHandler((text: string) => {
    convertValue(primaryField, text)
      .then(amount => {
        if (amount != null) {
          const otherField = Number(!primaryField)
          const newAmounts = []
          newAmounts[primaryField] = text
          newAmounts[otherField] = amount
          setAmounts(newAmounts)
        }
      })
      .catch(e => console.log(e.message))
  })

  const bottomRow = useHandler((fieldNum: number) => {
    const primaryAmount = amounts[fieldNum]
    const amountBlank = eq(primaryAmount, '0')
    const currencyNameStyle = amountBlank ? styles.bottomCurrencyMuted : styles.bottomCurrency
    const currencyName = fieldInfos[fieldNum].currencyName

    return (
      <View style={styles.bottomContainer} key="bottom">
        <View style={styles.valueContainer}>
          <NumericInput
            style={styles.bottomAmount}
            value={primaryAmount}
            maxDecimals={fieldInfos[fieldNum].maxEntryDecimals}
            placeholder={s.strings.string_amount}
            placeholderTextColor={theme.deactivatedText}
            onChangeText={onNumericInputChange}
            autoCorrect={false}
            returnKeyType={topReturnKeyType}
            ref={inputRefs[fieldNum]}
            onSubmitEditing={onNext}
            inputAccessoryViewID={inputAccessoryViewID}
          />
          <EdgeText style={currencyNameStyle}>{' ' + currencyName}</EdgeText>
        </View>
      </View>
    )
  })

  const topRow = useHandler((fieldNum: number) => {
    let topText = amounts[fieldNum]
    if (isValidInput(topText)) {
      topText = formatNumberInput(topText, { minDecimals: 0, maxDecimals: fieldInfos[fieldNum].maxEntryDecimals })
    }

    const fieldInfo = fieldInfos[fieldNum]
    topText = `${topText} ${fieldInfo.currencyName}`
    return (
      <TouchableWithoutFeedback onPress={onToggleFlipInput} key="top">
        <EdgeText>{topText}</EdgeText>
      </TouchableWithoutFeedback>
    )
  })

  if (startAmounts.length !== 2 || fieldInfos.length !== 2) throw new Error('Invalid number of startAmounts or fieldInfos')

  useEffect(() => {
    if (getMethods != null)
      getMethods({
        setAmounts: amounts => {
          setAmounts([amounts[0], amounts[1]])
        }
      })
  }, [])

  return (
    <>
      <View style={styles.flipInputContainer}>
        <View style={styles.flipInput}>
          <Animated.View style={[styles.flipInputFront, frontAnimatedStyle]} pointerEvents={!primaryField ? 'auto' : 'none'}>
            {topRow(1)}
            {bottomRow(0)}
          </Animated.View>
          <Animated.View style={[styles.flipInputFront, styles.flipContainerBack, backAnimatedStyle]} pointerEvents={primaryField ? 'auto' : 'none'}>
            {topRow(0)}
            {bottomRow(1)}
          </Animated.View>
        </View>
        <ButtonBox onPress={onToggleFlipInput} paddingRem={[0.5, 0, 0.5, 1]}>
          <Fontello style={styles.flipIcon} name="exchange" color={theme.iconTappable} size={theme.rem(1.5)} />
        </ButtonBox>
      </View>
    </>
  )
})

const getStyles = cacheStyles((theme: Theme) => ({
  // Header
  headerContainer: {
    marginRight: Platform.OS === 'ios' ? theme.rem(3.5) : theme.rem(1.5), // Different because adjustsFontSizeToFit behaves differently on android vs ios
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: theme.rem(1)
  },
  headerText: {
    fontWeight: '600',
    fontSize: theme.rem(1.0)
  },

  // Flip Input
  flipInputContainer: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  flipInput: {
    flex: 1,
    paddingRight: theme.rem(0.5)
  },
  flipInputFront: {
    backfaceVisibility: 'hidden'
  },
  flipContainerBack: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0
  },
  flipIcon: {
    marginRight: -theme.rem(0.125)
  },

  // Top Amount
  bottomContainer: {
    flexDirection: 'row',
    marginRight: theme.rem(1.5),
    minHeight: theme.rem(2)
  },
  valueContainer: {
    flexDirection: 'row',
    marginRight: theme.rem(0.5)
  },
  bottomAmount: {
    color: theme.primaryText,
    includeFontPadding: false,
    fontFamily: theme.fontFaceMedium,
    fontSize: theme.rem(1.5),
    minHeight: theme.rem(1.5)
  },
  bottomCurrency: {
    paddingTop: theme.rem(0.125)
  },
  bottomCurrencyMuted: {
    paddingTop: theme.rem(0.125),
    color: theme.deactivatedText
  }
}))
