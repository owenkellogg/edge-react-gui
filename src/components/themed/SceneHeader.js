// @flow

import * as React from 'react'
import { StyleSheet, View } from 'react-native'

import { memo } from '../../types/reactHooks.js'
import { Space } from '../layout/Space'
import { type Theme, cacheStyles, useTheme } from '../services/ThemeContext.js'
import { DividerLine } from './DividerLine.js'
import { EdgeText } from './EdgeText.js'

type Props = {
  title?: string,
  children?: React.Node,
  icon?: React.Node,
  underline?: boolean,
  withTopMargin?: boolean,
  style?: StyleSheet.Styles
}

export const SceneHeaderComponent = (props: Props) => {
  const { title, underline, withTopMargin, children, icon = null, style } = props
  const theme = useTheme()
  const styles = getStyles(theme)
  return (
    <>
      <View style={[styles.container, withTopMargin ? styles.topMargin : null, style]}>
        {title ? <EdgeText style={styles.title}>{title}</EdgeText> : null}
        {children}
        {icon ? <Space right>{icon}</Space> : null}
      </View>
      <View style={styles.dividerLine}>{underline ? <DividerLine /> : null}</View>
    </>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    marginLeft: theme.rem(1),
    paddingBottom: theme.rem(1),
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center'
  },
  topMargin: {
    marginTop: theme.rem(1)
  },
  dividerLine: {
    marginLeft: theme.rem(1),
    marginBottom: theme.rem(0.5)
  },
  title: {
    fontSize: theme.rem(1.2),
    fontFamily: theme.fontFaceMedium
  }
}))

export const SceneHeader = memo(SceneHeaderComponent)
