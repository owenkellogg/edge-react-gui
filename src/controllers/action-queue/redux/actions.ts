import { Dispatch, GetState } from '../../../types/reduxTypes'
import { ActionQueueStore, makeActionQueueStore } from '../ActionQueueStore'
import { uploadPushEvents } from '../push'
import { getEffectPushEventIds } from '../runtime'
import { ActionProgram, ActionProgramState, ActionQueueItem, ActionQueueMap } from '../types'

type LoadActionQueueStateAction = {
  type: 'ACTION_QUEUE/LOAD_QUEUE'
  data: ActionQueueMap
}

type ScheduleProgramAction = {
  type: 'ACTION_QUEUE/QUEUE_ITEM'
  programId: string
  item: ActionQueueItem
}

type UpdateProgramStateAction = {
  type: 'ACTION_QUEUE/UPDATE_PROGRAM_STATE'
  state: ActionProgramState
}

export type ActionQueueAction = LoadActionQueueStateAction | ScheduleProgramAction | UpdateProgramStateAction

export const scheduleActionProgram = (program: ActionProgram) => async (dispatch: Dispatch, getState: GetState) => {
  const state = getState()
  const clientId = state.core.context.clientId
  const store: ActionQueueStore = makeActionQueueStore(state.core.account, clientId)
  const programId = program.programId

  // Persist the ActionProgram to the ActionQueueStore
  const programState = await store.createActionQueueItem(program)

  dispatch({
    type: 'ACTION_QUEUE/QUEUE_ITEM',
    programId,
    item: {
      program,
      state: programState
    }
  })
}

export const updateActionProgramState = (programState: ActionProgramState) => async (dispatch: Dispatch, getState: GetState) => {
  const state = getState()
  const clientId = state.core.context.clientId
  const store: ActionQueueStore = makeActionQueueStore(state.core.account, clientId)

  await store.updateActionQueueItem(programState)

  dispatch({ type: 'ACTION_QUEUE/UPDATE_PROGRAM_STATE', state: programState })
}

export const cancelActionProgram = (programId: string) => async (dispatch: Dispatch, getState: GetState) => {
  const state = getState()
  const account = state.core.account
  const clientId = state.core.context.clientId
  const store: ActionQueueStore = makeActionQueueStore(state.core.account, clientId)
  const { state: programState } = state.actionQueue.actionQueueMap[programId]

  const pushEventIds = getEffectPushEventIds(programState.effect)
  if (pushEventIds.length > 0) {
    await uploadPushEvents({ account, clientId }, { removeEvents: pushEventIds })
  }

  programState.effect = { type: 'done', cancelled: true }

  await store.updateActionQueueItem(programState)

  dispatch({ type: 'ACTION_QUEUE/UPDATE_PROGRAM_STATE', state: programState })
}
