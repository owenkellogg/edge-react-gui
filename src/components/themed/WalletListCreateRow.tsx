import { EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import { View } from 'react-native'
import { TouchableOpacity } from 'react-native-gesture-handler'

import { createWallet, CreateWalletOptions } from '../../actions/CreateWalletActions'
import { approveTokenTerms } from '../../actions/TokenTermsActions'
import { showFullScreenSpinner } from '../../components/modals/AirshipFullScreenSpinner'
import { Airship, showError } from '../../components/services/AirshipInstance'
import { getSpecialCurrencyInfo } from '../../constants/WalletAndCurrencyConstants'
import { useHandler } from '../../hooks/useHandler'
import { useWatch } from '../../hooks/useWatch'
import s from '../../locales/strings'
import { useDispatch, useSelector } from '../../types/reactRedux'
import { Dispatch, GetState } from '../../types/reduxTypes'
import { getCreateWalletType } from '../../util/CurrencyInfoHelpers'
import { CryptoIcon } from '../icons/CryptoIcon'
import { ListModal } from '../modals/ListModal'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { EdgeText } from './EdgeText'
import { WalletListCurrencyRow } from './WalletListCurrencyRow'

export type WalletListCreateRowProps = {
  currencyCode: string
  currencyName: string
  createWalletIds?: string[]
  pluginId?: string
  walletType?: string

  onPress?: (walletId: string, currencyCode: string) => void
}

export const WalletListCreateRowComponent = (props: WalletListCreateRowProps) => {
  const {
    currencyCode = '',
    currencyName = '',
    createWalletIds = [],
    walletType,
    pluginId,

    // Callbacks:
    onPress
  } = props
  const account = useSelector(state => state.core.account)
  const currencyWallets = useWatch(account, 'currencyWallets')

  const dispatch = useDispatch()
  const theme = useTheme()
  const styles = getStyles(theme)

  const handlePress = useHandler(() => {
    // @ts-expect-error
    const handleRes = walletId => (onPress != null ? onPress(walletId, currencyCode) : null)
    if (walletType != null) {
      dispatch(createAndSelectWallet({ walletType })).then(handleRes)
    } else if (pluginId != null) {
      if (createWalletIds.length < 2) {
        dispatch(createAndSelectToken({ tokenCode: currencyCode, pluginId, createWalletId: createWalletIds[0] })).then(handleRes)
      } else {
        Airship.show(bridge => {
          const renderRow = (wallet: EdgeCurrencyWallet) => (
            <WalletListCurrencyRow
              wallet={wallet}
              onPress={walletId => {
                dispatch(createAndSelectToken({ tokenCode: currencyCode, pluginId: currencyWallets[walletId].currencyInfo.pluginId, createWalletId: walletId }))
                  .then(handleRes)
                  .finally(() => bridge.resolve())
              }}
            />
          )

          return (
            <ListModal<EdgeCurrencyWallet>
              bridge={bridge}
              title={s.strings.select_wallet}
              textInput={false}
              fullScreen={false}
              rowComponent={renderRow}
              rowsData={createWalletIds.map(walletId => currencyWallets[walletId])}
            />
          )
        })
      }
    }
  })

  return (
    <TouchableOpacity style={styles.row} onPress={handlePress}>
      <CryptoIcon currencyCode={currencyCode} marginRem={1} pluginId={pluginId} sizeRem={2} />
      <View style={styles.nameColumn}>
        <EdgeText style={styles.currencyText}>{currencyCode}</EdgeText>
        <EdgeText style={styles.nameText}>{currencyName}</EdgeText>
      </View>
      <View style={styles.labelColumn}>
        <EdgeText style={styles.labelText}>{walletType != null ? s.strings.fragment_create_wallet_create_wallet : s.strings.wallet_list_add_token}</EdgeText>
      </View>
    </TouchableOpacity>
  )
}

const createAndSelectToken =
  ({ tokenCode, pluginId, createWalletId }: { tokenCode: string; pluginId: string; createWalletId?: string }) =>
  async (dispatch: Dispatch, getState: GetState): Promise<string> => {
    const state = getState()
    const { account, disklet } = state.core
    const { defaultIsoFiat } = state.ui.settings
    const parentCurrencyCode = account.currencyConfig[pluginId].currencyInfo.currencyCode

    try {
      // Show the user the token terms modal only once
      await approveTokenTerms(disklet, parentCurrencyCode)

      // Try to find existing Parent Edge Wallet, if no specific wallet was given
      const { currencyWallets } = account
      const parentWalletId =
        createWalletId ?? Object.keys(currencyWallets).find(walletId => currencyWallets[walletId].currencyInfo.currencyCode === parentCurrencyCode)
      const wallet: EdgeCurrencyWallet =
        parentWalletId != null
          ? currencyWallets[parentWalletId]
          : // If no parent chain wallet exists, create it
            await showFullScreenSpinner(
              s.strings.wallet_list_modal_enabling_token,
              (async (): Promise<EdgeCurrencyWallet> => {
                const { walletType } = getCreateWalletType(account, parentCurrencyCode) ?? {}
                if (walletType == null) throw new Error(s.strings.create_wallet_failed_message)
                return await createWallet(account, {
                  walletType,
                  walletName: getSpecialCurrencyInfo(walletType).initWalletName,
                  fiatCurrencyCode: defaultIsoFiat
                })
              })()
            )
      await wallet.enableTokens([tokenCode])
      return wallet.id
    } catch (error: any) {
      showError(error)
    }
    return ''
  }

const createAndSelectWallet = ({ walletType, fiatCurrencyCode, walletName }: CreateWalletOptions) => {
  walletName = walletName ?? getSpecialCurrencyInfo(walletType).initWalletName
  return async (dispatch: Dispatch, getState: GetState) => {
    const state = getState()
    const { account } = state.core
    try {
      const wallet = await showFullScreenSpinner(
        s.strings.wallet_list_modal_creating_wallet,
        createWallet(account, { walletName, walletType, fiatCurrencyCode })
      )
      return wallet.id
    } catch (error: any) {
      showError(error)
    }
    return ''
  }
}

/**
 * Renders a WalletListRow with "Create Wallet" children
 * */
const getStyles = cacheStyles((theme: Theme) => ({
  // Layout:
  row: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'center',
    marginLeft: theme.rem(0.5),
    minHeight: theme.rem(4.25)
  },
  labelColumn: {
    justifyContent: 'center',
    paddingRight: theme.rem(1)
  },
  nameColumn: {
    flexDirection: 'column',
    flexGrow: 1,
    flexShrink: 1,
    marginRight: theme.rem(0.5)
  },

  // Text:
  currencyText: {
    flexBasis: 'auto',
    flexShrink: 1,
    fontFamily: theme.fontFaceMedium
  },
  labelText: {
    fontFamily: theme.fontFaceMedium
  },
  nameText: {
    fontSize: theme.rem(0.75),
    color: theme.secondaryText
  }
}))

export const WalletListCreateRow = React.memo(WalletListCreateRowComponent)
