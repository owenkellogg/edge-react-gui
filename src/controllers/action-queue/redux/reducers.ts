import { combineReducers, Reducer } from 'redux'

import { Action } from '../../../types/reduxTypes'
import { checkEffectIsDone } from '../runtime'
import { ActionQueueMap } from '../types'

export type ActionQueueState = {
  readonly actionQueueMap: ActionQueueMap
}

export const actionQueue: Reducer<ActionQueueState, Action> = combineReducers({
  actionQueueMap(state: ActionQueueMap = {}, action: Action): ActionQueueMap {
    switch (action.type) {
      case 'ACTION_QUEUE/LOAD_QUEUE': {
        return action.data
      }
      case 'ACTION_QUEUE/QUEUE_ITEM': {
        return {
          ...state,
          [action.programId]: action.item
        }
      }
      case 'ACTION_QUEUE/UPDATE_PROGRAM_STATE': {
        const { programId } = action.state
        const { effect } = action.state

        // Remove program from action queue if it has finished
        if (checkEffectIsDone(effect)) {
          const { [programId]: _removed, ...rest } = state
          return rest
        }

        return {
          ...state,
          [programId]: {
            ...state[programId],
            state: action.state
          }
        }
      }
      case 'LOGOUT': {
        return {}
      }
      default:
        return state
    }
  }
})
