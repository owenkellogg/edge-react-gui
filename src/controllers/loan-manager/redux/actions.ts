import { EdgeAccount } from 'edge-core-js'

import { BorrowPlugin } from '../../../plugins/borrow-plugins/types'
import { Dispatch, GetState } from '../../../types/reduxTypes'
import { makeCleanStore } from '../../../util/CleanStore'
import { scheduleActionProgram } from '../../action-queue/redux/actions'
import { ActionProgram } from '../../action-queue/types'
import { borrowPluginMap } from '../borrowPluginConfig'
import { asLoanAccountMapRecord, LOAN_ACCOUNT_MAP, LOAN_MANAGER_STORE_ID, LoanAccountEntry, LoanProgramEdge, LoanProgramType } from '../store'
import { LoanAccount } from '../types'
import { checkLoanHasFunds } from '../util/checkLoanHasFunds'

type SetLoanAccountAction = {
  type: 'LOAN_MANAGER/SET_LOAN_ACCOUNT'
  loanAccount: LoanAccount
}
type DeleteLoanAccountAction = {
  type: 'LOAN_MANAGER/DELETE_LOAN_ACCOUNT'
  id: string
}
type UpdateSyncRatio = {
  type: 'LOAN_MANAGER/SET_SYNC_RATIO'
  syncRatio: number
}

export type LoanManagerActions = SetLoanAccountAction | DeleteLoanAccountAction | UpdateSyncRatio

/**
 * Save a new LoanAccount to disk and sets it in the Redux state.
 * It will throw if specified LoanAccount exists already.
 */
export const createLoanAccount = (loanAccount: LoanAccount) => async (dispatch: Dispatch, getState: GetState) => {
  const state = getState()
  const account: EdgeAccount = state.core.account
  const store = makeCleanStore(account, LOAN_MANAGER_STORE_ID)
  const loanAccountMapRecord = await store.initRecord(LOAN_ACCOUNT_MAP, asLoanAccountMapRecord)

  if (loanAccountMapRecord.data[loanAccount.id] == null) {
    const { borrowPlugin, borrowEngine, closed, programEdges } = loanAccount
    const loanEntry: LoanAccountEntry = {
      closed,
      walletId: borrowEngine.currencyWallet.id,
      borrowPluginId: borrowPlugin.borrowInfo.borrowPluginId,
      programEdges
    }

    loanAccountMapRecord.update({ ...loanAccountMapRecord.data, [loanAccount.id]: loanEntry })
  } else {
    throw new Error('Creating duplicate LoanAccount id: ' + loanAccount.id)
  }

  dispatch({
    type: 'LOAN_MANAGER/SET_LOAN_ACCOUNT',
    loanAccount: loanAccount
  })
}

/**
 * Load all LoanAccounts from disk to the redux store and initializes
 * associated BorrowEngines.
 */
export const loadLoanAccounts = (account: EdgeAccount) => async (dispatch: Dispatch, getState: GetState) => {
  const store = makeCleanStore(account, LOAN_MANAGER_STORE_ID)
  const loanAccountMapRecord = await store.initRecord(LOAN_ACCOUNT_MAP, asLoanAccountMapRecord)

  for (const key of Object.keys(loanAccountMapRecord.data)) {
    const loanAccountEntry = loanAccountMapRecord.data[key]
    const { walletId, borrowPluginId, closed, programEdges } = loanAccountEntry
    const wallet = await account.waitForCurrencyWallet(walletId)
    const borrowPlugin = borrowPluginMap[borrowPluginId]
    const borrowEngine = await borrowPlugin.makeBorrowEngine(wallet)
    const loanAccount: LoanAccount = {
      id: walletId,
      borrowPlugin,
      borrowEngine,
      closed,
      programEdges
    }
    dispatch({
      type: 'LOAN_MANAGER/SET_LOAN_ACCOUNT',
      loanAccount
    })
  }
}

/**
 * Update an existing LoanAccount.
 * It will throw if specified LoanAccount is not found.
 */
export const updateLoanAccount = (loanAccount: LoanAccount) => async (dispatch: Dispatch, getState: GetState) => {
  const state = getState()
  const account: EdgeAccount = state.core.account
  const store = makeCleanStore(account, LOAN_MANAGER_STORE_ID)
  const loanAccountMapRecord = await store.initRecord(LOAN_ACCOUNT_MAP, asLoanAccountMapRecord)

  // Create loan if it doesn't exist
  if (loanAccountMapRecord.data[loanAccount.id] == null) {
    dispatch(await createLoanAccount(loanAccount))
    return
  }

  const { closed, programEdges } = loanAccount
  loanAccountMapRecord.data[loanAccount.id].closed = closed
  loanAccountMapRecord.data[loanAccount.id].programEdges = programEdges
  loanAccountMapRecord.update(loanAccountMapRecord.data)

  dispatch({
    type: 'LOAN_MANAGER/SET_LOAN_ACCOUNT',
    loanAccount: loanAccount
  })
}

/**
 * Delete an existing LoanAccount.
 * It will throw if specified LoanAccount is not found.
 */
export const deleteLoanAccount = (loanAccountOrId: LoanAccount | string) => async (dispatch: Dispatch, getState: GetState) => {
  const state = getState()
  const account: EdgeAccount = state.core.account
  const store = makeCleanStore(account, LOAN_MANAGER_STORE_ID)
  const loanAccountMapRecord = await store.initRecord(LOAN_ACCOUNT_MAP, asLoanAccountMapRecord)

  const loanAccountId = typeof loanAccountOrId === 'string' ? loanAccountOrId : loanAccountOrId.id
  if (loanAccountMapRecord.data[loanAccountId] != null) {
    const { [loanAccountId]: _, ...loanAccountMapRecordUpdated } = loanAccountMapRecord.data
    loanAccountMapRecord.update(loanAccountMapRecordUpdated)
  } else {
    throw new Error('Could not find LoanAccount id: ' + loanAccountId)
  }

  dispatch({
    type: 'LOAN_MANAGER/DELETE_LOAN_ACCOUNT',
    id: loanAccountId
  })
}

/**
 * Execute action program, and update associated LoanAccount.
 */
export const runLoanActionProgram =
  (loanAccount: LoanAccount, actionProgram: ActionProgram, programType: LoanProgramType) =>
  async (dispatch: Dispatch, getState: GetState): Promise<LoanAccount> => {
    await dispatch(scheduleActionProgram(actionProgram))
    const programEdge: LoanProgramEdge = {
      programId: actionProgram.programId,
      programType
    }
    await dispatch(updateLoanAccount({ ...loanAccount, programEdges: [...loanAccount.programEdges, programEdge] }))

    return loanAccount
  }

export const resyncLoanAccounts = (account: EdgeAccount) => async (dispatch: Dispatch, getState: GetState) => {
  const store = makeCleanStore(account, LOAN_MANAGER_STORE_ID)
  const loanAccountMapRecord = await store.initRecord(LOAN_ACCOUNT_MAP, asLoanAccountMapRecord)

  const typeHack: any = Object.values(borrowPluginMap)
  const borrowPlugins: BorrowPlugin[] = typeHack

  // `loanAccountIds` is synonymous to `walletIds`
  const ids = Object.keys(account.currencyWallets)

  dispatch({
    type: 'LOAN_MANAGER/SET_SYNC_RATIO',
    syncRatio: 0
  })

  let progress = 0
  const promises = ids
    .map(async (id, i) => {
      const loanAccountId = id
      try {
        const wallet = await account.waitForCurrencyWallet(loanAccountId)

        const currencyPluginId = wallet.currencyConfig.currencyInfo.pluginId
        const borrowPlugin = borrowPlugins.find(borrowPlugin => borrowPlugin.borrowInfo.currencyPluginId === currencyPluginId)

        if (borrowPlugin == null) return

        const borrowEngine = await borrowPlugin.makeBorrowEngine(wallet)

        await new Promise(resolve => {
          borrowEngine.watch('syncRatio', syncRatio => {
            // @ts-expect-error
            if (syncRatio >= 1) resolve()
          })
        })

        if (checkLoanHasFunds(borrowEngine)) {
          if (loanAccountMapRecord.data[loanAccountId]) return

          // Create a loan account for the wallet
          const loanAccount: LoanAccount = {
            id: loanAccountId,
            borrowPlugin,
            borrowEngine,
            closed: false,
            programEdges: []
          }
          await dispatch(createLoanAccount(loanAccount))
        } else {
          const existingLoanAccount = loanAccountMapRecord.data[loanAccountId]
          if (existingLoanAccount != null && existingLoanAccount.closed) {
            await dispatch(deleteLoanAccount(loanAccountId))
          }
        }
      } catch (error: any) {
        console.error(error)
      }
    })
    .map(async promise =>
      promise.then(() => {
        const syncRatio = ++progress / ids.length
        // console.log('\n###\n', { total: ids.length, progress, syncRatio }, '\n###\n')
        dispatch({
          type: 'LOAN_MANAGER/SET_SYNC_RATIO',
          syncRatio
        })
      })
    )

  await Promise.all(promises)

  dispatch({
    type: 'LOAN_MANAGER/SET_SYNC_RATIO',
    syncRatio: 1
  })
}
