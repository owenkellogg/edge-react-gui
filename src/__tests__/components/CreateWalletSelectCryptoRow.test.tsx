import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import IonIcon from 'react-native-vector-icons/Ionicons'
import { Provider } from 'react-redux'
import { createRenderer } from 'react-test-renderer/shallow'
import { createStore } from 'redux'

import { CreateWalletSelectCryptoRow } from '../../components/themed/CreateWalletSelectCryptoRow'
import { rootReducer } from '../../reducers/RootReducer'

describe('WalletListRow', () => {
  const mockState: any = {
    core: {
      account: {
        currencyConfigs: {
          bitcoin: {
            pluginId: 'bitcoin'
          }
        }
      }
    }
  }
  const store = createStore(rootReducer, mockState)

  it('should render with loading props', () => {
    const renderer = createRenderer()

    const props = {
      pluginId: 'bitcoin',
      walletName: 'My bitcoin wallet',
      onPress: () => undefined
    }
    const actual = renderer.render(
      <Provider store={store}>
        <CreateWalletSelectCryptoRow {...props}>
          <IonIcon size={26} color="#66EDA8" name="chevron-forward-outline" />
        </CreateWalletSelectCryptoRow>
      </Provider>
    )

    expect(actual).toMatchSnapshot()
  })
})
