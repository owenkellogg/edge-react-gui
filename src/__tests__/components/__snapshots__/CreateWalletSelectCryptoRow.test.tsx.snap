// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletListRow should render with loading props 1`] = `
<Context.Provider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Memo(CreateWalletSelectCryptoRowComponent)
    onPress={[Function]}
    pluginId="bitcoin"
    walletName="My bitcoin wallet"
  >
    <Icon
      allowFontScaling={false}
      color="#66EDA8"
      name="chevron-forward-outline"
      size={26}
    />
  </Memo(CreateWalletSelectCryptoRowComponent)>
</Context.Provider>
`;
